examples/cfq/syntax-based/node.py

```diff
@@ -0,0 +1,119 @@
+from asg import Action, APPLY_RULE, GENERATE_TOKEN
+
+class Node:
+
+  def __init__(self, parent: 'Node', value: str):
+    self.parent = parent
+    self.value = value
+    self.children = []
+
+  def add_child(self, child: 'Node'):
+    self.children.append(child)
+
+
+def apply_action(frontier_nodes_stack: deque, action: Action, grammar: Grammar):
+  """Applies an action (apply rule or generate token). The action extends a
+  AST that is under construction by extending the stack of frontier nodes.
+  The function returns the extended stack."""
+  current_node: Node = frontier_nodes_stack.pop()
+  action_type, action_value = action
+  if action_type == GENERATE_TOKEN:
+    # Generate leaf node with token stored in the action.
+    child = Node(current_node, action_value)
+    current_node.add_child(child)
+  else:
+    new_frontier_nodes = []
+    rule_name = action_value
+    head, body = grammar.sub_rules[rule_name]
+    if head != current_node.value:
+      raise Exception('Invalid action. Got {} and expected {}'.format(
+          head, current_node.value))
+    rule_tokens = body.split()
+    for rule_token in rule_tokens:
+      match = re.match(r'\"(.*)\"', rule_token)
+      if match:
+        # Create a leaf node with a token from the rule (e.g. SELECT).
+        node_value = match.groups()[0]
+        child = Node(current_node, node_value)
+        current_node.add_child(child)
+      else:
+        # Create new frontier node.
+        child = Node(current_node, rule_token)
+        current_node.add_child(child)
+        new_frontier_nodes.append(child)
+    new_frontier_nodes.reverse()
+    frontier_nodes_stack.extend(new_frontier_nodes)
+  return frontier_nodes_stack
+
+
+def apply_first_action(action: Action, grammar: Grammar):
+  """Applies the first action in the sequence of actions and constructs the
+  tree root."""
+  action_type, action_value = action
+  if action_type != APPLY_RULE:
+    raise Exception('First action should be rule application')
+  _, node_value = grammar.sub_rules[action_value]
+  root = Node(None, node_value)
+  return root
+
+
+def apply_sequence_of_actions(action_sequence: List, grammar: Grammar):
+  """Applies a sequence of actions to construct a syntax tree."""
+  root = apply_first_action(action_sequence[0], grammar)
+  frontier_nodes = deque()
+  frontier_nodes.append(root)
+  for action in action_sequence[1:]:
+    frontier_nodes = apply_action(frontier_nodes, action, grammar)
+  return root
+
+
+def traverse_tree(root: Node):
+  """DFS raversal of tree. The result of the traversal should be the query. In
+  case the parent node is a terminal, the descendents will be simply
+  concatenated, e.g. VAR, otherwise the substrings are merged together by spaces.
+  """
+  if len(root.children) == 0:
+    return root.value
+  children_substrings = []
+  for child in root.children:
+    child_substr = traverse_tree(child)
+    children_substrings.append(child_substr)
+  if root.value.isupper():
```

Comment:
Is this only to make sure "?x" is concatenated with an integer, or are there other cases as well?

If so, perhaps it is easier to explicitly add all "?x<num>"'s to the vocabulary. I don't think there are so many right?

Reply:
What do you mean adding them to the vocabulary?

Reply:
Grammar was simplified to not break variable in "?x" and digit